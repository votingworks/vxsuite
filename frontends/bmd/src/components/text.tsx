import React from 'react';
import styled, { css } from 'styled-components';
import DomPurify from 'dompurify';

import * as GLOBALS from '../config/globals';

interface Props {
  bold?: boolean;
  light?: boolean;
  center?: boolean;
  error?: boolean;
  italic?: boolean;
  muted?: boolean;
  narrow?: boolean;
  noWrap?: boolean;
  small?: boolean;
  warning?: boolean;
  warningIcon?: boolean;
  wordBreak?: boolean;
  voteIcon?: boolean;
}

const iconStyles = css<Props>`
  &::before {
    display: inline-block;
    margin-top: -0.3rem;
    margin-right: 0.25rem;
    border-radius: ${({ warningIcon }) => warningIcon && '50%'};
    background: ${({ warningIcon, voteIcon }) =>
      (warningIcon && 'darkorange') || (voteIcon && '#028099')};
    width: 1rem;
    height: 1rem;
    vertical-align: middle;
    text-align: center;
    line-height: 1.1;
    color: #ffffff;
    font-size: 90%;
    font-weight: 800;
    content: ${({ warningIcon, voteIcon }) =>
      (warningIcon && "'!'") || (voteIcon && `'${GLOBALS.CHECK_ICON}'`)};
  }
`;

export const Text = styled('p')<Props>`
  margin-right: ${({ narrow }) => (narrow ? 'auto' : undefined)};
  margin-left: ${({ narrow }) => (narrow ? 'auto' : undefined)};
  max-width: ${({ narrow }) => (narrow ? '33ch' : undefined)};
  text-align: ${({ center }) => (center ? 'center' : undefined)};
  white-space: ${({ noWrap }) => (noWrap ? 'nowrap' : undefined)};
  color: ${({ error, muted, warning }) =>
    (error && 'red') ??
    (warning && 'darkorange') ??
    (muted && 'gray') ??
    undefined};
  @media print {
    color: ${({ error, muted, warning }) =>
      (error && 'black') ??
      (warning && 'black') ??
      (muted && 'black') ??
      undefined};
  }
  font-size: ${({ small }) => (small ? '0.8rem' : undefined)};
  font-weight: ${({ bold, light }) =>
    (bold && '600') ?? (light && '300') ?? undefined};
  font-style: ${({ italic }) => (italic ? 'italic' : undefined)};
  word-break: ${({ wordBreak }) => (wordBreak ? 'break-word' : undefined)};
  /* stylelint-disable-next-line value-keyword-case, order/properties-order */
  ${({ warningIcon, voteIcon }) => (warningIcon || voteIcon) && iconStyles}
`;

interface TextWithLineBreaksProps extends Props {
  text: string;
  style?: React.CSSProperties;
}

export function TextWithLineBreaks({
  text,
  ...rest
}: TextWithLineBreaksProps): JSX.Element {
  return (
    <React.Fragment>
      {text.split(/[\n\r]{2}/g).map((x) => (
        <Text {...rest} key={x}>
          {x.split(/[\n\r]/g).map((y, i, arr) => (
            <React.Fragment key={y}>
              <span
                // eslint-disable-next-line react/no-danger
                dangerouslySetInnerHTML={{
                  __html: DomPurify.sanitize(y),
                }}
              />
              {i !== arr.length - 1 && <br />}
            </React.Fragment>
          ))}
        </Text>
      ))}
    </React.Fragment>
  );
}

export const NoWrap = styled.span`
  white-space: nowrap;
`;
