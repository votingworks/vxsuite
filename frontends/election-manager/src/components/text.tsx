import React from 'react';
import styled, { css } from 'styled-components';

import * as GLOBALS from '../config/globals';

interface Props {
  bold?: boolean;
  light?: boolean;
  center?: boolean;
  error?: boolean;
  italic?: boolean;
  left?: boolean;
  muted?: boolean;
  narrow?: boolean;
  normal?: boolean;
  noWrap?: boolean;
  preLine?: boolean;
  right?: boolean;
  small?: boolean;
  warning?: boolean;
  warningIcon?: boolean;
  wordBreak?: boolean;
  voteIcon?: boolean;
  white?: boolean;
}

const iconStyles = css<Props>`
  &::before {
    display: inline-block;
    margin-top: -0.3em;
    margin-right: 0.25em;
    border-radius: ${({ warningIcon }) => warningIcon && '50%'};
    background: ${({ warningIcon, voteIcon }) =>
      (warningIcon && 'darkorange') || (voteIcon && '#028099')};
    width: 1em;
    height: 1em;
    vertical-align: middle;
    text-align: center;
    line-height: 1.1;
    color: #ffffff;
    font-size: 90%;
    font-weight: 800;
    content: ${({ warningIcon, voteIcon }) =>
      (warningIcon && "'!'") || (voteIcon && `'${GLOBALS.CHECK_ICON}'`)};
  }
`;

export const Text = styled('p')<Props>`
  margin-right: ${({ narrow }) => (narrow ? 'auto' : undefined)};
  margin-left: ${({ narrow }) => (narrow ? 'auto' : undefined)};
  max-width: ${({ narrow }) => (narrow ? '33ch' : undefined)};
  text-align: ${({ center, right, left }) =>
    (left && 'left') ||
    (center && 'center') ||
    (right && 'right') ||
    undefined};
  white-space: ${({ noWrap, preLine }) =>
    noWrap ? 'nowrap' : preLine ? 'pre-line' : undefined};
  color: ${({ error, muted, warning, white }) =>
    (error && 'red') ??
    (warning && 'darkorange') ??
    (muted && 'gray') ??
    (white && '#FFFFFF') ??
    undefined};
  @media print {
    color: ${({ error, muted, warning, white }) =>
      (error && 'black') ??
      (warning && 'black') ??
      (muted && 'black') ??
      (white && '#FFFFFF') ??
      undefined};
  }
  font-size: ${({ small }) => (small ? '0.8em' : undefined)};
  font-weight: ${({ bold, light, normal }) =>
    (bold && '600') ?? (light && '300') ?? (normal && '400') ?? undefined};
  font-style: ${({ italic }) => (italic ? 'italic' : undefined)};
  word-break: ${({ wordBreak }) => (wordBreak ? 'break-word' : undefined)};
  /* stylelint-disable-next-line value-keyword-case, order/properties-order */
  ${({ warningIcon, voteIcon }) => (warningIcon || voteIcon) && iconStyles}
`;

export function TextWithLineBreaks({ text }: { text: string }): JSX.Element {
  return (
    <React.Fragment>
      {text.split(/[\n\r]{2}/g).map((x) => (
        <p key={x}>
          {x.split(/[\n\r]/g).map((y, i, arr) => (
            <React.Fragment key={y}>
              {y}
              {i !== arr.length - 1 && <br />}
            </React.Fragment>
          ))}
        </p>
      ))}
    </React.Fragment>
  );
}

export const NoWrap = styled.span`
  white-space: nowrap;
`;

export const Monospace = styled.span`
  font-family: monospace;
`;
