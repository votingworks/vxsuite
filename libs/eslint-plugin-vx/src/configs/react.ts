import recommended = require('./recommended');

const jsExtensions = ['.js', '.jsx'];
const tsExtensions = ['.ts', '.tsx'];
const allExtensions = jsExtensions.concat(tsExtensions);

export = {
  ...recommended,
  extends: [
    ...recommended.extends
      .map((name) => (name === 'airbnb-base' ? 'airbnb' : name))
      .filter((name) => name !== 'plugin:prettier/recommended'),
    'airbnb/hooks',
    'plugin:react/recommended',
    'plugin:jsx-a11y/recommended',
    'plugin:prettier/recommended', // Enables eslint-plugin-prettier and displays prettier errors as ESLint errors. Make sure this is always the last configuration in the extends array.
  ],
  settings: {
    ...recommended.settings,
    react: {
      version: 'detect',
    },
  },
  rules: {
    ...recommended.rules,
    'jsx-a11y/label-has-associated-control': [
      'error',
      {
        controlComponents: ['Select'],
      },
    ],
    'react/jsx-filename-extension': ['error', { extensions: allExtensions }],
    'react/jsx-fragments': ['error', 'element'],
    'react/jsx-no-bind': 'off',
    'react/jsx-no-constructed-context-values': 'off',
    'react/jsx-no-useless-fragment': ['error', { allowExpressions: true }],
    'react/jsx-one-expression-per-line': 'off',
    'react/jsx-props-no-spreading': 'off',
    'react/jsx-wrap-multilines': 'off',
    'react/prop-types': 'off',
    'react/require-default-props': 'off',
  },
};
