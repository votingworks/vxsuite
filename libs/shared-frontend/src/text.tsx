import React from 'react';
import styled, { css } from 'styled-components';
import DomPurify from 'dompurify';

import * as GLOBALS from './globals';

export const successTextGreen = 'rgb(0,128,0)';

interface Props {
  bold?: boolean;
  light?: boolean;
  center?: boolean;
  error?: boolean;
  success?: boolean;
  italic?: boolean;
  left?: boolean;
  muted?: boolean;
  narrow?: boolean;
  normal?: boolean;
  noWrap?: boolean;
  preLine?: boolean;
  right?: boolean;
  small?: boolean;
  warning?: boolean;
  warningIcon?: boolean;
  wordBreak?: boolean;
  voteIcon?: boolean;
  white?: boolean;
}

const iconStyles = css<Props>`
  &::before {
    display: inline-block;
    margin-top: -0.3em;
    margin-right: 0.25em;
    border-radius: ${({ warningIcon }) => warningIcon && '50%'};
    background: ${({ warningIcon, voteIcon }) =>
      (warningIcon && 'darkorange') || (voteIcon && '#028099')};
    width: 1em;
    height: 1em;
    vertical-align: middle;
    text-align: center;
    line-height: 1.1;
    color: #ffffff;
    font-size: 90%;
    font-weight: 800;
    content: ${({ warningIcon, voteIcon }) =>
      (warningIcon && "'!'") || (voteIcon && `'${GLOBALS.CHECK_ICON}'`)};
  }
`;

export const Text = styled('p')<Props>`
  margin-right: ${({ narrow }) => (narrow ? 'auto' : undefined)};
  margin-left: ${({ narrow }) => (narrow ? 'auto' : undefined)};
  max-width: ${({ narrow }) => (narrow ? '33ch' : undefined)};
  text-align: ${({ center, right, left }) =>
    (left && 'left') ||
    (center && 'center') ||
    (right && 'right') ||
    undefined};
  white-space: ${({ noWrap, preLine }) =>
    noWrap ? 'nowrap' : preLine ? 'pre-line' : undefined};
  color: ${({ error, muted, success, warning, white }) =>
    (error && 'red') ??
    (warning && 'darkorange') ??
    (success && successTextGreen) ??
    (muted && 'gray') ??
    (white && '#FFFFFF') ??
    undefined};
  @media print {
    color: ${({ error, muted, success, warning, white }) =>
      (error && 'black') ??
      (warning && 'black') ??
      (success && 'black') ??
      (muted && 'black') ??
      (white && '#FFFFFF') ??
      undefined};
  }
  font-size: ${({ small }) => (small ? '0.8em' : undefined)};
  font-weight: ${({ bold, light, normal }) =>
    (bold && '600') ?? (light && '300') ?? (normal && '400') ?? undefined};
  font-style: ${({ italic }) => (italic ? 'italic' : undefined)};
  word-break: ${({ wordBreak }) => (wordBreak ? 'break-word' : undefined)};
  /* stylelint-disable-next-line value-keyword-case, order/properties-order */
  ${({ warningIcon, voteIcon }) => (warningIcon || voteIcon) && iconStyles}
`;

interface TextWithLineBreaksProps extends Props {
  text: string;
  style?: React.CSSProperties;
}

export function TextWithLineBreaks({
  text,
  ...rest
}: TextWithLineBreaksProps): JSX.Element {
  return (
    <React.Fragment>
      {text.split(/[\n\r]{2}/g).map((x) => (
        <Text {...rest} key={x}>
          {x.split(/[\n\r]/g).map((y, i, arr) => (
            <React.Fragment key={y}>
              <span
                // eslint-disable-next-line react/no-danger
                dangerouslySetInnerHTML={{
                  __html: DomPurify.sanitize(y),
                }}
              />
              {i !== arr.length - 1 && <br />}
            </React.Fragment>
          ))}
        </Text>
      ))}
    </React.Fragment>
  );
}

export const NoWrap = styled.span`
  white-space: nowrap;
`;

export const Monospace = styled.span`
  font-family: monospace;
`;
