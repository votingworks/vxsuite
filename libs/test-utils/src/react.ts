/* eslint-disable no-console -- this file monkey-patches `console` */

/**
 * Suppresses the "Can't perform a React state update on an unmounted component" warning by
 * modifying the `console.error` method to ignore the warning.
 *
 * This is a temporary workaround until we upgrade to React 18.
 */
export function suppressReact17UnmountedWarning(): void {
  const originalConsoleError = console.error;
  console.error = (...args: unknown[]) => {
    if (
      typeof args[0] === 'string' &&
      args[0].startsWith(
        `Warning: Can't perform a React state update on an unmounted component`
      )
    ) {
      return;
    }

    originalConsoleError(...args);
  };
}
