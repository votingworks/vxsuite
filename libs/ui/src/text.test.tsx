import React from 'react';
import { render } from '@testing-library/react';
import 'jest-styled-components';

import { Text, TextWithLineBreaks } from './text';

describe('renders Text', () => {
  test('as paragraph tag', async () => {
    const text = 'paragraph';
    const { getByText } = render(<Text>{text}</Text>);
    const element = getByText(text);
    expect(element.tagName).toEqual('P');
    expect(element).toMatchSnapshot();
  });

  test('center muted', async () => {
    const { container } = render(
      <Text center muted>
        center muted?
      </Text>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  test('displays error style', async () => {
    const { container } = render(<Text error>Error Text?</Text>);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('narrow wordBreak warning', async () => {
    const { container } = render(
      <Text narrow wordBreak warning>
        narrow wordBreak
      </Text>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  test('vote icon', async () => {
    const { container } = render(<Text voteIcon>vote!</Text>);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('warning icon', async () => {
    const { container } = render(<Text warningIcon>Warning</Text>);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('warning icon/vote icon toggle', async () => {
    const toggle = true;
    let { container } = render(
      <Text warningIcon={toggle} voteIcon={!toggle}>
        Warning
      </Text>
    );
    expect(container.firstChild).toHaveStyleRule('content', "'!'", {
      modifier: '::before',
    });
    expect(container.firstChild).toHaveStyleRule('background', 'darkorange', {
      modifier: '::before',
    });
    expect(container.firstChild).toHaveStyleRule('border-radius', '50%', {
      modifier: '::before',
    });

    ({ container } = render(
      <Text warningIcon={!toggle} voteIcon={toggle}>
        Success
      </Text>
    ));
    expect(container.firstChild).toHaveStyleRule('content', "'✓'", {
      modifier: '::before',
    });
    expect(container.firstChild).toHaveStyleRule('background', '#028099', {
      modifier: '::before',
    });
  });

  test('align left preLine normal italic', async () => {
    const { container } = render(
      <Text left preLine normal italic>
        align left preLine
      </Text>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  test('align right nowrap light', async () => {
    const { container } = render(
      <Text right noWrap light>
        align right nowrap
      </Text>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  test('small white bold', async () => {
    const { container } = render(
      <Text small white bold>
        small white bold
      </Text>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  test('converts line-breaks into <p> and <br/> tags', () => {
    const { container } = render(
      <TextWithLineBreaks
        text={'I’m a paragraph.\n\nAnd I’m a paragraph with a\nline break.'}
      />
    );
    expect(container).toMatchInlineSnapshot(`
      <div>
        <p>
          I’m a paragraph.
        </p>
        <p>
          And I’m a paragraph with a
          <br />
          line break.
        </p>
      </div>
    `);
  });
});
