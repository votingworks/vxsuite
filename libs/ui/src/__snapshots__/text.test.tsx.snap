// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Text align left preLine normal italic 1`] = `
.c0 {
  text-align: left;
  white-space: pre-line;
  font-weight: 400;
  font-style: italic;
}

@media print {

}

<p
  class="c0"
>
  align left preLine
</p>
`;

exports[`renders Text align right nowrap light 1`] = `
.c0 {
  text-align: right;
  white-space: nowrap;
  font-weight: 300;
}

@media print {

}

<p
  class="c0"
>
  align right nowrap
</p>
`;

exports[`renders Text as paragraph tag 1`] = `
@media print {

}

<p
  class=""
>
  paragraph
</p>
`;

exports[`renders Text center muted 1`] = `
.c0 {
  text-align: center;
  color: gray;
}

@media print {
  .c0 {
    color: black;
  }
}

<p
  class="c0"
>
  center muted?
</p>
`;

exports[`renders Text displays error style 1`] = `
.c0 {
  color: red;
}

@media print {
  .c0 {
    color: black;
  }
}

<p
  class="c0"
>
  Error Text?
</p>
`;

exports[`renders Text narrow wordBreak warning 1`] = `
.c0 {
  margin-right: auto;
  margin-left: auto;
  max-width: 33ch;
  color: darkorange;
  word-break: break-word;
}

@media print {
  .c0 {
    color: black;
  }
}

<p
  class="c0"
>
  narrow wordBreak
</p>
`;

exports[`renders Text small white bold 1`] = `
.c0 {
  color: #FFFFFF;
  font-size: 0.8em;
  font-weight: 600;
}

@media print {
  .c0 {
    color: #FFFFFF;
  }
}

<p
  class="c0"
>
  small white bold
</p>
`;

exports[`renders Text vote icon 1`] = `
.c0::before {
  display: inline-block;
  margin-top: -0.3em;
  margin-right: 0.25em;
  background: #028099;
  width: 1em;
  height: 1em;
  vertical-align: middle;
  text-align: center;
  line-height: 1.1;
  color: #ffffff;
  font-size: 90%;
  font-weight: 800;
  content: 'âœ“';
}

@media print {

}

<p
  class="c0"
>
  vote!
</p>
`;

exports[`renders Text warning icon 1`] = `
.c0::before {
  display: inline-block;
  margin-top: -0.3em;
  margin-right: 0.25em;
  border-radius: 50%;
  background: darkorange;
  width: 1em;
  height: 1em;
  vertical-align: middle;
  text-align: center;
  line-height: 1.1;
  color: #ffffff;
  font-size: 90%;
  font-weight: 800;
  content: '!';
}

@media print {

}

<p
  class="c0"
>
  Warning
</p>
`;
